### PR TITLE
Add departments

### PR DESCRIPTION
**Pull request by Finemas**
Added downloading, saving and view for departments. I also added downloading document types and subtypes depending on the department.

### CODE REVIEW

File: zScanner/Services/Storage/Model/DocumentDatabaseModel.swift [hunk 1]
```diff
@@ -47,12 +47,12 @@ extension DocumentDatabaseModel {
...
             id: id,
             folder: folder!.toDomainModel(),
-            type: type!.toDomainModel(),
+            type: type?.toDomainModel() ?? DocumentTypeDomainModel(id: "", name: ""),
             date: date,
             name: name,
             notes: notes,
             pages: pages.map({ $0.toDomainModel() }),
-            department: department!.toDomainModel()
+            department: department?.toDomainModel() ?? DepartmentDomainModel(id: "", name: "")
```
Comment (skorepak) [inline zScanner/Services/Storage/Model/DocumentDatabaseModel.swift]: I'm just curious, why did you make this change?

File: zScanner/SupportFiles/Resources/cs.lproj/Localizable.strings [hunk 2]
```diff
@@ -34,6 +34,9 @@
 "document.modeSelector.cancel" = "Zrušit";
 "document.modeSelector.title" = "Zvolte typ záznamu";
 "document.emptyView.title" = "Nemáte žádné nahrané dokumenty, dokument vytvoříte klepnutím na tlačítko plus.";
+"departments.tableHeader" = "Pracoviště";
+"departments.fetchError" = "Chyba komunikace se serverem.";
+"departments.reloadButton" = "Zkusit znovu";
```
Comment (skorepak) [inline zScanner/SupportFiles/Resources/cs.lproj/Localizable.strings]: add `.title` to all of them

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 3]
```diff
@@ -72,62 +74,125 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
...
+        documentsViewModel.fetchDocumentTypes(for: departmentCode)
     }
     
     private func setupView() {
         navigationItem.title = "document.screen.title".localized
         
-        view.addSubview(tableView)
-        tableView.snp.makeConstraints { make in
-            make.edges.equalToSuperview()
+        documentsTableView.dataSource = self
```
Comment (skorepak) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: why did you move the datasource setup from the lazy init here? I think it should go back. Or is there any specific reason?

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 4]
```diff
@@ -72,62 +74,125 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
...
+        view.addSubview(documentsTableView)
+        documentsTableView.snp.makeConstraints { (make) in
```
Comment (skorepak) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: It's very cosmetic, but just for consistency, please remove the brackets around the `make`, they are not needed for single argument and it looks better :)

Comment (skorepak) [reply zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: check for `(make)` everywhere in the app, please

File: zScanner/ViewControllers/Documents/DocumentsListViewController.swift [hunk 5]
```diff
@@ -72,62 +74,125 @@ class DocumentsListViewController: BaseViewController, ErrorHandling {
...
+        
+        view.addSubview(departmentsContainerView)
+        departmentsContainerView.snp.makeConstraints { (make) in
+            make.top.equalTo(documentsTableView.snp.bottom).offset(16)
+            make.left.right.bottom.equalToSuperview()
+            make.height.equalTo(0).priority(100)
+            make.height.lessThanOrEqualTo(view.snp.height).multipliedBy(0.666)
+        }
+        
+        departmentsContainerView.backgroundColor = UIColor.lightGray.withAlphaComponent(0.3)
```
Comment (skorepak) [inline zScanner/ViewControllers/Documents/DocumentsListViewController.swift]: Please create lazy init for `departmentsContainerView` and move the background colour there.  I did it yesterday in a rush.

File: zScanner/ViewModels/Department/DepartmentsListViewModel.swift [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+class DepartmentsListViewModel {
+    enum DepartmentState {
+        case awaitingInteraction
+        case loading
+        case success
+        case error(RequestError)
+    }
+    
+    //MARK: Instance part
+    private let database: Database
```
Comment (skorepak) [inline zScanner/ViewModels/Department/DepartmentsListViewModel.swift]: database is not used, remove it

File: zScanner/ViewModels/Documents/DocumentsListViewModel.swift [hunk 7]
```diff
@@ -64,30 +60,38 @@ class DocumentsListViewModel {
...
             .map({ DocumentViewModel(document: $0.toDomainModel(), networkManager: networkManager, database: database) })
             .reversed()
     }
     
-    func fetchDocumentTypes() {
+    func fetchDocumentTypes(for departmentCode: String) {
+        if departmentCode == lastDepartmentSelected {
+            self.documentTypesState.onNext(.success)
+            return
+        }
```
Comment (skorepak) [inline zScanner/ViewModels/Documents/DocumentsListViewModel.swift:76]: 👍

**Review by skorepak — CHANGES_REQUESTED** [review]
Few comments added

File: zScanner/Services/Storage/Model/DocumentDatabaseModel.swift (hunk 1 above)
Comment (Finemas) [reply zScanner/Services/Storage/Model/DocumentDatabaseModel.swift]: That's a good point. I just realize that I shouldn't do it because each document has department and type. My bad.
The app sometimes falls down because of it but I should not happen, am I right?

File: zScanner/Views/Departments/DepartmentButton.swift [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    let disposedBag = DisposeBag()
+    
+    func setup() {
+        setTitle(model.name, for: .normal)
+        
+        rx.tap
+            .subscribe(onNext: {
```
Comment (skorepak) [inline zScanner/Views/Departments/DepartmentButton.swift]: weak self

File: zScanner/Services/Storage/Model/DocumentDatabaseModel.swift (hunk 1 above)
Comment (skorepak) [reply zScanner/Services/Storage/Model/DocumentDatabaseModel.swift]: It should not happen. It might be because of some data inconsistency during development, it should not happen in production

**Comment by skorepak** [comment]
One more comment

**Review by skorepak — COMMENTED** [review]
.

**Review by skorepak — APPROVED** [review]
